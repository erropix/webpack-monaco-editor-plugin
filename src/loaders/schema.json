{
    "title": "Monaco Editor Loader Options",
    "type": "object",
    "additionalProperties": false,
    "required": ["workers", "features", "languages"],
    "properties": {
        "workers": {
            "type": "object",
            "description": "Object containing a map of workers label to their entry file.",
            "additionalProperties": {
                "type": "string"
            }
        },
        "features": {
            "type": "array",
            "description": "Array of Monaco editor features entries to include.",
            "items": {
                "type": "string"
            }
        },
        "languages": {
            "type": "array",
            "description": "Array of Monaco editor languages entries to include.",
            "items": {
                "type": "string"
            }
        },
        "publicPath": {
            "type": "string",
            "description": "The public path to load the Monaco Editor from. This is used to load the editor from a CDN or a custom location."
        },
        "global": {
            "type": "boolean"
        }
    }
}
