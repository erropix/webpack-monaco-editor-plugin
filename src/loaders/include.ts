import schema from './schema.json';

import type { PitchLoaderDefinitionFunction } from 'webpack';
import type { Schema } from 'schema-utils';

export interface IMonacoEditorLoaderOptions {
    workers: Record<string, string>;
    features: string[];
    languages: string[];
    publicPath?: string;
    global?: boolean;
}

export const pitch: PitchLoaderDefinitionFunction<IMonacoEditorLoaderOptions> = function pitch(request: string) {
    const { workers, features, languages, publicPath = '', global = false } = this.getOptions(schema as Schema);

    const contextify = (entry: string) => {
        return this.utils.contextify(this.context, entry);
    };

    const buffer: string[] = [];

    buffer.push(`// Setup Monaco Environment to load workers
const base = ${JSON.stringify(publicPath)} || __webpack_public_path__;
const paths = ${JSON.stringify(workers, null, 4)};

function getWorkerUrl(id, label) {
    if (!paths.hasOwnProperty(label))
        throw new Error(\`The \${label} worker was not found. Supported workers: \${Object.keys(paths).join(', ')}\`);
    const path = paths[label];
    const url = new URL(path, new URL(base, location.origin));
    if (url.origin === location.origin)
        return url;
    return URL.createObjectURL(new Blob([\`import '\${url}';\`], { type: 'text/javascript' }));
}

self.MonacoEnvironment = {
    globalAPI: ${global},
    getWorkerUrl,
};`);
    buffer.push('');
    buffer.push('// Import feature modules');
    buffer.push(...features.map((entry) => `import "${contextify(entry)}";`));
    buffer.push('');
    buffer.push(`// Import Monaco Editor core`);
    buffer.push(`import * as monaco from "${contextify(`!!${request}`)}";`);
    buffer.push('');
    buffer.push('// Import language modules');
    buffer.push(...languages.map((entry) => `import "${contextify(entry)}";`));
    buffer.push('');
    buffer.push('// Export Monaco Editor API');
    buffer.push(`export * from "${contextify(`!!${request}`)}";`);
    buffer.push(`export default monaco;`);

    return buffer.join('\n');
};
