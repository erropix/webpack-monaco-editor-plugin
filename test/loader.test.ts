import type { Stats, StatsModule } from 'webpack';
import type { IMonacoEditorPluginOptions } from '../src/index';

import { describe, it, expect, beforeAll } from '@jest/globals';
import { pitch } from '../src/loaders/include';
import { compile } from './helpers';

function mockLoaderContext(options: any): any {
    return {
        getOptions: () => options,
        context: '/project',
        utils: {
            contextify: (_context: string, entry: string) => entry,
        },
    };
}

describe('MonacoEditorLoader pitch function', () => {
    it('should use default publicPath and global when not provided', () => {
        const ctx = mockLoaderContext({
            workers: {},
            features: [],
            languages: [],
            // publicPath and global omitted
        });
        const result = pitch.call(ctx, 'monaco-editor/esm/vs/editor/editor.api', '', {});

        expect(result).toContain('const base = "" || __webpack_public_path__');
        expect(result).toContain('globalAPI: false');
    });

    it('should use provided publicPath and global', () => {
        const ctx = mockLoaderContext({
            workers: {},
            features: [],
            languages: [],
            publicPath: '/static/',
            global: true,
        });
        const result = pitch.call(ctx, 'monaco-editor/esm/vs/editor/editor.api', '', {});

        expect(result).toContain('const base = "/static/" || __webpack_public_path__');
        expect(result).toContain('globalAPI: true');
    });
});

describe('MonacoEditorLoader', () => {
    let stats: Stats;
    let modules: StatsModule[];

    beforeAll(async () => {
        stats = await compile({
            webpack: {
                module: {
                    noParse: /[\\/]monaco-editor[\\/]esm[\\/]vs/,
                },
            },
        });

        modules = stats.toJson({ source: true }).modules || [];
    }, 10000);

    it('should generate correct Monaco entry source', () => {
        const module = modules[0];
        expect(module.source).toMatchSnapshot('monaco-editor-api-javascript-css-json-html-php');
    });
});
