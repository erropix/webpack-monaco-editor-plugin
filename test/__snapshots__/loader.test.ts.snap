// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonacoEditorLoader should generate correct Monaco entry source: monaco-editor-api-javascript-css-json-html-php 1`] = `
"// Setup Monaco Environment to load workers
const base = "" || __webpack_public_path__;
const paths = {
    "editorWorkerService": "editor.worker.js",
    "css": "css.worker.js",
    "html": "html.worker.js",
    "json": "json.worker.js",
    "typescript": "ts.worker.js",
    "javascript": "ts.worker.js"
};

function getWorkerUrl(id, label) {
    if (!paths.hasOwnProperty(label))
        throw new Error(\`The \${label} worker was not found. Supported workers: \${Object.keys(paths).join(', ')}\`);
    const path = paths[label];
    const url = new URL(path, new URL(base, location.origin));
    if (url.origin === location.origin)
        return url;
    return URL.createObjectURL(new Blob([\`import '\${url}';\`], { type: 'text/javascript' }));
}

self.MonacoEnvironment = {
    globalAPI: false,
    getWorkerUrl,
};

// Import feature modules
import "./contrib/anchorSelect/browser/anchorSelect.js";
import "./contrib/bracketMatching/browser/bracketMatching.js";
import "./browser/coreCommands.js";
import "./contrib/caretOperations/browser/caretOperations.js";
import "./contrib/caretOperations/browser/transpose.js";
import "./contrib/clipboard/browser/clipboard.js";
import "./contrib/codeAction/browser/codeActionContributions.js";
import "./browser/widget/codeEditor/codeEditorWidget.js";
import "./contrib/codelens/browser/codelensController.js";
import "./contrib/colorPicker/browser/colorContributions.js";
import "./contrib/colorPicker/browser/standaloneColorPickerActions.js";
import "./contrib/comment/browser/comment.js";
import "./contrib/contextmenu/browser/contextmenu.js";
import "./contrib/cursorUndo/browser/cursorUndo.js";
import "./browser/widget/diffEditor/diffEditor.contribution.js";
import "./contrib/diffEditorBreadcrumbs/browser/contribution.js";
import "./contrib/dnd/browser/dnd.js";
import "./contrib/documentSymbols/browser/documentSymbols.js";
import "./contrib/dropOrPasteInto/browser/copyPasteContribution.js";
import "./contrib/dropOrPasteInto/browser/dropIntoEditorContribution.js";
import "./contrib/find/browser/findController.js";
import "./contrib/folding/browser/folding.js";
import "./contrib/fontZoom/browser/fontZoom.js";
import "./contrib/format/browser/formatActions.js";
import "./contrib/gotoError/browser/gotoError.js";
import "./standalone/browser/quickAccess/standaloneGotoLineQuickAccess.js";
import "./contrib/gotoSymbol/browser/goToCommands.js";
import "./contrib/gotoSymbol/browser/link/goToDefinitionAtPosition.js";
import "./contrib/hover/browser/hoverContribution.js";
import "./standalone/browser/iPadShowKeyboard/iPadShowKeyboard.js";
import "./contrib/inPlaceReplace/browser/inPlaceReplace.js";
import "./contrib/indentation/browser/indentation.js";
import "./contrib/inlayHints/browser/inlayHintsContribution.js";
import "./contrib/inlineCompletions/browser/inlineCompletions.contribution.js";
import "./contrib/inlineEdit/browser/inlineEdit.contribution.js";
import "./contrib/inlineEdits/browser/inlineEdits.contribution.js";
import "./contrib/inlineProgress/browser/inlineProgress.js";
import "./standalone/browser/inspectTokens/inspectTokens.js";
import "./contrib/lineSelection/browser/lineSelection.js";
import "./contrib/linesOperations/browser/linesOperations.js";
import "./contrib/linkedEditing/browser/linkedEditing.js";
import "./contrib/links/browser/links.js";
import "./contrib/longLinesHelper/browser/longLinesHelper.js";
import "./contrib/multicursor/browser/multicursor.js";
import "./contrib/parameterHints/browser/parameterHints.js";
import "./contrib/placeholderText/browser/placeholderText.contribution.js";
import "./standalone/browser/quickAccess/standaloneCommandsQuickAccess.js";
import "./standalone/browser/quickAccess/standaloneHelpQuickAccess.js";
import "./standalone/browser/quickAccess/standaloneGotoSymbolQuickAccess.js";
import "./contrib/readOnlyMessage/browser/contribution.js";
import "./standalone/browser/referenceSearch/standaloneReferenceSearch.js";
import "./contrib/rename/browser/rename.js";
import "./contrib/sectionHeaders/browser/sectionHeaders.js";
import "./contrib/semanticTokens/browser/documentSemanticTokens.js";
import "./contrib/semanticTokens/browser/viewportSemanticTokens.js";
import "./contrib/smartSelect/browser/smartSelect.js";
import "./contrib/snippet/browser/snippetController2.js";
import "./contrib/stickyScroll/browser/stickyScrollContribution.js";
import "./contrib/suggest/browser/suggestController.js";
import "./contrib/suggest/browser/suggestInlineCompletions.js";
import "./standalone/browser/toggleHighContrast/toggleHighContrast.js";
import "./contrib/toggleTabFocusMode/browser/toggleTabFocusMode.js";
import "./contrib/tokenization/browser/tokenization.js";
import "./contrib/unicodeHighlighter/browser/unicodeHighlighter.js";
import "./contrib/unusualLineTerminators/browser/unusualLineTerminators.js";
import "./contrib/wordHighlighter/browser/wordHighlighter.js";
import "./contrib/wordOperations/browser/wordOperations.js";
import "./contrib/wordPartOperations/browser/wordPartOperations.js";

// Import Monaco Editor core
import * as monaco from "!!./editor.api.js";

// Import language modules
import "../basic-languages/css/css.contribution.js";
import "../language/css/monaco.contribution.js";
import "../basic-languages/html/html.contribution.js";
import "../language/html/monaco.contribution.js";
import "../basic-languages/javascript/javascript.contribution.js";
import "../language/json/monaco.contribution.js";
import "../basic-languages/php/php.contribution.js";
import "../basic-languages/typescript/typescript.contribution.js";
import "../language/typescript/monaco.contribution.js";

// Export Monaco Editor API
export * from "!!./editor.api.js";
export default monaco;"
`;
